File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-1.py
class Solution:
    from collections import defaultdict
    def characterReplacement(self, s: str, k: int) -> int:
        # init useful structs
        n = len(s)
        counts = defaultdict(int)
        L,res = 0,0

        # advance right pointer
        for R in range(n):
            counts[s[R]] +=1
            # while failing condition
            while sum(counts.values()) - max(counts.values()) > k:
                # advance left pointer
                counts[s[L]] -=1
                L+=1
            # update max valid
            res = max(res,R-L+1)

        return res

File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-1.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char,int> window;
        int res = 0;
        int N = s.size();

        int L = 0, R = 0;
        while (R < N) {
            while (window[s[R]]) {
                window[s[L]]--;
                L++;
            }
            window[s[R]]++;
            res = max(res,R-L+1);
            R++;
        }

        return res;
    }
};
